xpra/server/auth/keycloak.py

```diff
@@ -158,9 +164,36 @@ def check(self, response_json: bytes) -> bool:
                 log.error("Error: keycloak authentication failed as token state is invalid")
                 return False
 
+            log("keycloak: token is active")
+
             user_info = keycloak_openid.userinfo(access_token)
-            log("userinfo_info: %r", user_info)
-            log("keycloak authentication succeeded: token is active")
+            log("user_info: %r", user_info)
+
+            log("claim_field: %r", self.claim_field)
+            log("auth_groups: %r", self.auth_groups)
+            # if claim_field is defined, check that the auth_groups match the groups_claim condition
+            if self.claim_field is not None and self.claim_field:
+              if not self.auth_groups or self.auth_groups is None:
+                log.error("Error: keycloak authentication failed as auth_groups is invalid")
+                return False
+              else:
+                def get(dict, keys):
+                  return get(dict[keys.pop(0)], keys) if len(keys) else dict
```

Comment:
What does this recursive function do?
It is confusing.

Reply:
This is because groups claim usually come nested in `user_info`. For example: 
```
"resource_access": {
   "roles": {
     "group": [
       "client-group-1",
       "client-group-2"
     ]
   }
}
```
The get function will go through the dictionary keys following `resource_access.roles.group` defined in `claim_field` until it finds the groups and return that. This way of defining the `claim_field` is used in other implementations for example in Nextcloud, so I defined it here in the same way.

Reply:
At least rename the argument, `dict` is a type.
And add a newline before the function - or just run the git commit hooks to validate everything.